Add item requests

Zazergel opened the pull request.

Zazergel committed:
ShareIt v2.3 add-bookings
feat: Добавлена функциональность добавления комментариев, бронирования вещей.
feat: Данные теперь хранятся в БД, а не в памяти.

Zazergel committed:
ShareIt v2.3.1 add-bookings
Попытка исправить ошибку из target #1

Zazergel committed:
ShareIt v2.3.2 add-bookings
Попытка исправить ошибку из target #2

Zazergel committed:
ShareIt v2.4.0 add-bookings
refactor: Рефакторинг мапперов ItemMapper, BookingMapper
del: Удалены классы MapperService и MapperServiceImpl
feat: Добавлен класс StateException
fix: Исправлены аннотации OneToOne на ManyToOne в классах: Comment,  Item, Booking
del: Удалены лишние поля в DTO классах: BookingRequestDto, CommentRequestDto

Zazergel committed:
ShareIt v2.5.0
fix: Исправлена проблема N+1 в методе getByOwnerId(Long userId) класса ItemServiceImpl
feat: добавлена проверка пересечений по датам для новых бронирований в методе add(Long userId) класса BookingServiceImpl

Zazergel committed:
ShareIt v2.5.1
fix: исправлены ошибки codeStyle

Zazergel committed:
ShareIt v2.6
feat: Реализованы реквесты
feat: Реализована пагинация
feat: Добавлены тесты
feat: Реализован Searcher, который выполняет поиск бронирований по bookerId или по ownerId в зависимости от переданного State

Zazergel committed:
ShareIt v2.6.1
fix: fix codestyle in ItemRepository

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,68 @@
+    @GetMapping
+    public List<BookingResponseDto> getAllByBookerId(
+            @RequestHeader(UserController.headerUserId) Long userId,
+            @RequestParam(defaultValue = "ALL", required = false) String state,
```
required не нужен, если задано дефолтное значение
Аналогично для остальных контроллеров

NikolaiDem replied:
В методе getAllByOwnerId() не исправлено

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/chainSearcher/booker/ChainSearcherByBooker.java:
```diff
@@ -0,0 +1,24 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+public class ChainSearcherByBooker {
```
Для поиска бронирований цепочка обязанностей не выглядит очень подходящей, потому что
1) Все обработчики находятся на одном логическом уровне
2) Порядок обработки не имеет значения
3) Обработать запрос может максимум один обработчик

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+        if (!item.getAvailable()) {
+            throw new BookingException("Предмет недоступен для бронирования.");
+        }
+        if (!bookingRepository.findByItemIdAndStartBeforeAndEndAfterAndStatusEqualsOrderByStartAsc(item.getId(),
```
Можно проще:
        if (savedBookings.stream()
                .filter(b -> Status.APPROVED.equals(b.getStatus()))
                .anyMatch(b -> b.getStart().isBefore(bookingRequestDto.getEnd()) && b.getEnd().isAfter(bookingRequestDto.getStart()))) {
            throw new BookingException("Предмет недоступен для бронирования. В это время его еще кто-то использует!");
        }

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -0,0 +1,60 @@
+    @OneToMany(fetch = FetchType.LAZY)
+    @OnDelete(action = OnDeleteAction.CASCADE)
+    @JoinColumn(name = "ITEM_ID", referencedColumnName = "ID")
+    List<Comment> comments;
```
Поле лишнее

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/Booking.java:
```diff
@@ -1,21 +1,60 @@
+    @Column(name = "END_DATE", nullable = false)
+    LocalDateTime end;
+
+    @ManyToOne
```
Лучше сделать Lazy инициализацию, чтобы уменьшить нагрузку на БД (N+1 issue)
https://vladmihalcea.com/n-plus-1-query-problem/
Аналогично для всех entity

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,80 +2,234 @@
+        }
+    }
+
+    private List<CommentDto> addComment(Item item) {
```
Не используется

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/ItemRequest.java:
```diff
@@ -1,17 +1,56 @@
+
+    @OneToMany
+    @JoinColumn(name = "REQUEST_ID", referencedColumnName = "ID")
+    List<Item> items;
```
Поле лишнее, N+1 issue

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,68 @@
+    private final BookingService bookingService;
+
+    @GetMapping("/{id}")
+    public BookingResponseDto getById(@RequestHeader(UserController.headerUserId) Long userId,
```
headerUserId лучше вынести в класс с общими константами

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,97 @@
+        return itemRequests.stream()
+                .map((itemRequest) -> itemRequestMapper.toItemRequestExtendedDto(
+                        itemRequest,
+                        itemRequest.getItems()
```
N+1 query issue, все предметы нужно получить для всего списка запросов одним селектом )
Аналогично для getByRequesterId

NikolaiDem reviewed (changes requested):
100 измененных классов - это прекрасно )

Zazergel committed:
ShareIt v2.6.2
feat: Класс с общими константами
fix: after 1 review (fix N+1) in ItemRequestServiceImpl
delete: ChainSearcher

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+
+        List<ItemRequestExtendedDto> result = itemRequests
+                .stream()
+                .filter(itemRequest -> itemRequest.getRequesterId().getId().equals(userId))
```
Для чего фильтр, если в запросе (71 строка) учитывается userId ?)
